# Making the most of RNA-seq experiments at CRUK-CI

High-throughput RNA-sequencing is now the standard technique for quantifying transcript abundance in a biological sample of interest. In this course we will describe the automated processes that place once you submit library to the Genomics Core for RNA sequencing, and what data you should expect to receive from the Bioinformatics Core.

We will describe the steps involved to go from sequencing library to a list of genes that show statistically significant differences between your biological conditions of interest. Practical sessions will focus on the interpretation of the data you receive from the Bioinformatics Core, and **not** how the pipeline is executed.

# Aims: During this course you will learn about

- Experimental design considerations for RNA-seq
- The Bioinformatics Core pipeline for RNA-seq analysis
- What data files you will receive back from the Bioinformatics Core after performing an RNA-seq experiment at CRUK-CI
- Galaxy and other online tools that can be used on the data you receive

# Objectives: After this course you should be able to

- Prepare for an Experimental Design meeting with Genomics and Bioinformatics Core
- Recall the key steps involved in the Bioinformatics RNA-seq pipeline
- Import the results into Galaxy (or similar) for further Biological interpretation
